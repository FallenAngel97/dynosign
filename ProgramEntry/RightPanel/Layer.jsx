import React from 'react';
import './Layer.scss';
import { connect } from 'react-redux';
import { changeActiveLayer, changeLayer, changeLayerVisibility } from '../actions'
import hiddenEye from './hidden_eye.svg';
import visibleEye from './visible_eye.svg';
import PropTypes from 'prop-types';

/**
 * Core module, which handles layer manipulation logic in the right panel
 * @module Layer
 */

export class Layer extends React.Component {

  state = {
    editLayer: false
  }
  
  _onKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.setState({ editLayer: false });
    }
  }
  showLayerOptions = (e) => {
    this.props.openContext(e);
  }
  UNSAFE_componentWillMount () {
    document.addEventListener('mousedown', this.handleClick, false);
  }
  DragOver = (e) => {
    this.props.layerDragOver(e);
  }
  ondragStart = (e) => {
    this.props.layerDragStart(e);
  }
  dragEnd = (e) => {
    this.props.layerDragEnd(e);
  }
  ondrop = (e) => {
    this.props.layerDrop(e);
  }
  dragLeave = (e) => {
    this.props.layerDragLeave(e);
  }
  componentWillUnmount () {
    document.removeEventListener('mousedown', this.handleClick, false);
  }
  changeVisibility = () => {
    const { layer } = this.props;
    this.props.changeLayerVisibility(!layer.hidden, this.props.layerId);
  }
  /**
   * Hides the text rename box, if clicked outside. If clicked on the layer single time,
   * changes the active layer to the clicked one
   * @param {MouseEvent} ev - neccessary to check, what is the clicked target
   */
  handleClick = (ev) => {
    if (!this.node.contains(ev.target)) {
      this.setState({ editLayer: false })
    } else if (ev.target.className.indexOf('singleLayer') > -1) {
      this.props._changeActiveLayer(this.props.layer, this.props.layerId);
    }
  }
  /**
   * Opens the rename text box
   */
  renameLayer = () => {
    this.setState({ editLayer: true });
  }
  layerNameChange = (ev) => {
    let { layer } = this.props.changeActiveLayer;
    layer.name = ev.target.value;
    this.props.changeLayer(layer, this.props.changeActiveLayer.layerId)
  }
  render () {
    const { editLayer } = this.state;
    const { changeActiveLayer, layerId, layer } = this.props;

    return (
      <div onDragStart={this.ondragStart} draggable={!editLayer}
        ref={node => { this.node = node }}
        onDoubleClick={this.renameLayer}
        onDrop={this.ondrop}
        onKeyPress={this._onKeyPress}
        onContextMenu={this.showLayerOptions}
        onDragLeave={this.dragLeave}
        onDragOver={this.DragOver}
        onDragEnd={this.dragEnd}
        className={'singleLayer' + ((changeActiveLayer.layerNumber === layerId) ? ' activeLayer' : '')}>
        <img onClick={this.changeVisibility} src={layer.hidden ? hiddenEye : visibleEye}/>
        {editLayer ? <input onChange={this.layerNameChange} type='text' className='singleLayerEdit' value={layer.name} /> : layer.name}
      </div>
    )
  }
}

Layer.propTypes = {
  layer: PropTypes.object,
  changeLayerVisibility: PropTypes.func,
  layerId: PropTypes.number,
  _changeActiveLayer: PropTypes.func,
  changeActiveLayer: PropTypes.object,
  changeLayer: PropTypes.func,
  layerDragStart: PropTypes.func,
  layerDragEnd: PropTypes.func,
  layerDrop: PropTypes.func,
  layerDragOver: PropTypes.func,
  layerDragLeave: PropTypes.func,
  openContext: PropTypes.func
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
  return {
    _changeActiveLayer: (layer, layerNumber) => dispatch(changeActiveLayer(layer, layerNumber)),
    changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
    changeLayerVisibility: (hidden, layerId) => dispatch(changeLayerVisibility(hidden, layerId))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Layer);
