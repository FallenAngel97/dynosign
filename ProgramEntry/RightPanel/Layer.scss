.activeLayer {
    background: rgba(139, 146, 99, 0.2);
    input {
        width: 159px;
    }
}
.singleLayer{
    transition: .2s ease-in-out padding;
    img {
        height: 11px;
    }
}
.drag_under {
    padding-top: 8px;
    padding-bottom: 8px;
    color: red;
}