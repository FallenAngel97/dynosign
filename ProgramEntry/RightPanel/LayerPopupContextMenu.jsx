import React from 'react';
import phoneLayout from '../MainArea/phone_layout.svg';
import desktopLayout from '../MainArea/desktop_layout.svg';
import './LayerPopupContextMenu.scss';
import PropTypes from 'prop-types';

export default class LayerPopupContextMenu extends React.Component {
  constructor (props) {
    super(props);
    this.dismissPopup = this.dismissPopup.bind(this);
  }
  componentDidMount () {
    document.addEventListener('click', this.dismissPopup, false)
  }
  componentWillUnmount () {
    document.removeEventListener('click', this.dismissPopup, false)
  }
  dismissPopup (e) {
    if (!this.node.contains(e.target)) {
      this.props.dismissPopup();
    }
  }
  render () {
    return (
      <div ref={ node => { this.node = node } } style={{ display: !this.props.hidden ? 'block' : 'none' }} id='layer_popup_contextmenu'>
        <div id='layout_context_selector'>
          <img src={phoneLayout} />
          <img src={desktopLayout} />
        </div>
        <ul>
          <li>Move up</li>
          <li>Move down</li>
          <li>Style</li>
        </ul>
      </div>
    )
  }
}
LayerPopupContextMenu.propTypes = {
  hidden: PropTypes.bool,
  dismissPopup: PropTypes.func
}
