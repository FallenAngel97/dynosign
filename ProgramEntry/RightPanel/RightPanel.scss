#rightPanel {
    width: 200px;
    grid-column: 3;
    display: flex;
    flex-direction: column;
}
#stackingLayers {
    display: flex;
    flex-direction: column-reverse;
    flex-grow: 1;
}
#layersButtonsBottom {
    margin-bottom: 40px;
}
