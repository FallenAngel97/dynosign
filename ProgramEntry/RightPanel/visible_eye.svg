<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   id="svg8"
   version="1.1"
   viewBox="0 0 179.91666 57.452305"
   height="57.452305mm"
   width="179.91666mm">
  <defs
     id="defs2" />
  <metadata
     id="metadata5">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" />
        <dc:title></dc:title>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g
     transform="translate(241.90453,-61.143185)"
     id="layer1">
    <path
       id="path815"
       d="m -151.9462,61.143185 c -31.73007,0 -89.95833,12.860853 -89.95833,28.725894 0,15.865031 58.22826,28.726411 89.95833,28.726411 31.73006,0 89.958329,-12.86137 89.958329,-28.726411 0,-15.86503 -58.228269,-28.725894 -89.958329,-28.725894 z m 0,12.094869 a 16.630953,16.630953 0 0 1 16.63102,16.631025 16.630953,16.630953 0 0 1 -16.63102,16.631031 16.630953,16.630953 0 0 1 -16.63103,-16.631031 16.630953,16.630953 0 0 1 16.63103,-16.631025 z"
       style="fill:#333333;fill-opacity:1;stroke:none;stroke-width:19.52677917;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1" />
    <circle
       r="6.8035712"
       cy="89.869339"
       cx="-151.9462"
       id="path833"
       style="fill:#333333;fill-opacity:1;stroke:none;stroke-width:19.52677917;stroke-miterlimit:4;stroke-dasharray:none;stroke-opacity:1" />
  </g>
</svg>
