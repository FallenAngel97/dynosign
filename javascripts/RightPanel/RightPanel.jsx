import React from 'react';
import './RightPanel.scss';
import { connect } from 'react-redux';
import { addLayer, deleteLayer, changeLayer, reorderLayers } from '../actions';
import Layer from './Layer';
import PropTypes from 'prop-types';

/**
 * The panel on the right side, including layers, opacity and buttons
 * @module RightPanel
 */

export class RightPanel extends React.Component {
  constructor (props) {
    super(props);
    this.changeLayerOpacity = this.changeLayerOpacity.bind(this);
    this.layerDragStart = this.layerDragStart.bind(this);
    this.layerDragEnd = this.layerDragEnd.bind(this);
    this.layerDrop = this.layerDrop.bind(this);
    this.layerDragOver = this.layerDragOver.bind(this);
    this.selectedLayer = null;
  }
  layerDragStart (e) {
    e.dataTransfer.effectAllowed = 'move'
    e.dataTransfer.setData('text/plain', null);
    this.selectedLayer = e.target;
  }
  layerDragEnd (e) {
  }
  layerDrop (e) {
    if (this.selectedLayer === e.target) return;
    const newIndex = [...e.target.parentNode.children].indexOf(e.target);
    const oldIndex = [...this.selectedLayer.parentNode.children].indexOf(this.selectedLayer);
    this.props.reorderLayers(oldIndex, newIndex);
  }
  layerDragOver (e) {
    if (e.preventDefault) {
      e.preventDefault(); // Necessary. Allows us to drop.
    }
    e.dataTransfer.dropEffect = 'move'; // See the section on the DataTransfer object.
    return false;
  }
  changeLayerOpacity (ev) {
    let layer = this.props.changeActiveLayer.layer;
    layer.opacity = ev.target.value;
    this.props.changeLayer(layer, this.props.changeActiveLayer.layerNumber)
  }
  render () {
    return (
      <div id='rightPanel'>
        <div id='topRightPanelIndicators'>
          <h3>Layers</h3>
          <input onChange={this.changeLayerOpacity} value={this.props.changeActiveLayer.layer.opacity} type="range" min="0" max="100" />
        </div>
        <div id='stackingLayers'>
          {this.props.layersCRUD.map((layer, index) => {
            return <Layer
              key={index} layerId={index}
              layerDragStart={this.layerDragStart}
              layerDragEnd={this.layerDragEnd}
              layerDrop={this.layerDrop}
              layerDragOver={this.layerDragOver}
              layer={layer} />
          })}
        </div>
        <div id='layersButtonsBottom'>
          <button>Group</button>
          <button onClick={() => this.props.addLayer(this.props.layersCRUD.length)}>+</button>
          <button disabled={this.props.layersCRUD.length === 1} onClick={() => this.props.deleteLayer(this.props.changeActiveLayer.layerNumber)}>-</button>
        </div>
      </div>
    )
  }
}

RightPanel.propTypes = {
  changeActiveLayer: PropTypes.object,
  layersCRUD: PropTypes.array,
  changeLayer: PropTypes.func,
  addLayer: PropTypes.func,
  deleteLayer: PropTypes.func,
  reorderLayers: PropTypes.func
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
  return {
    deleteLayer: (layerId) => dispatch(deleteLayer(layerId)),
    addLayer: (layersCount) => dispatch(addLayer(layersCount)),
    changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
    reorderLayers: (oldIndex, newIndex) => dispatch(reorderLayers(oldIndex, newIndex))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(RightPanel);
