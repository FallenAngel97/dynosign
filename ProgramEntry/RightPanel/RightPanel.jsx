import React from 'react';
import './RightPanel.scss';
import { connect } from 'react-redux';
import { addLayer, deleteLayer, changeLayer, reorderLayers } from '../actions';
import Layer from './Layer';
import PropTypes from 'prop-types';
import LayerPopupContextMenu from './LayerPopupContextMenu';
import ReactDOM from 'react-dom';
import { Translate } from 'react-localize-redux';

/**
 * The panel on the right side, including layers, opacity and buttons
 * @module RightPanel
 */

export class RightPanel extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      displayContext: false
    }
    this.changeLayerOpacity = this.changeLayerOpacity.bind(this);
    this.layerDragStart = this.layerDragStart.bind(this);
    this.layerDragEnd = this.layerDragEnd.bind(this);
    this.layerDrop = this.layerDrop.bind(this);
    this.layerDragOver = this.layerDragOver.bind(this);
    this.selectedLayer = null;
    this.openContext = this.openContext.bind(this);
    this.dismissPopup = this.dismissPopup.bind(this);
  }
  layerDragStart (e) {
    e.dataTransfer.effectAllowed = 'move'
    e.dataTransfer.setData('text/plain', null);
    this.selectedLayer = e.target;
  }
  layerDragEnd (e) {
  }
  layerDrop (e) {
    if (this.selectedLayer === e.target) return;
    e.target.className = 'singleLayer';
    const newIndex = [...e.target.parentNode.children].indexOf(e.target);
    const oldIndex = [...this.selectedLayer.parentNode.children].indexOf(this.selectedLayer);
    this.props.reorderLayers(oldIndex, newIndex);
  }
  layerDragOver (e) {
    if (e.preventDefault) {
      e.preventDefault(); // Necessary. Allows us to drop.
    }
    if (this.selectedLayer === e.target) return;
    e.target.className += ' drag_under';
    e.dataTransfer.dropEffect = 'move'; // See the section on the DataTransfer object.
    return false;
  }
  layerDragLeave (e) {
    e.target.className = 'singleLayer';
  }
  changeLayerOpacity (ev) {
    let layer = this.props.changeActiveLayer.layer;
    layer.opacity = ev.target.value;
    this.props.changeLayer(layer, this.props.changeActiveLayer.layerNumber)
  }
  dismissPopup () {
    this.setState({ displayContext: false });
  }
  openContext (e) {
    const elem = ReactDOM.findDOMNode(this.context);
    elem.style.left = e.clientX + 'px';
    elem.style.top = e.clientY + 'px';
    if (window.innerHeight - e.clientY < 145) {
      elem.style.top = e.clientY - 145 + 'px';
    }
    if (window.innerWidth - e.clientX < 106) {
      elem.style.left = e.clientX - 106 + 'px';
    }
    this.setState({ displayContext: true });
  }
  render () {
    return (
      <div id='rightPanel'>
        <div id='topRightPanelIndicators'>
          <h3><Translate id='layers'/></h3>
          <input onChange={this.changeLayerOpacity} value={this.props.changeActiveLayer.layer.opacity} type="range" min="0" max="100" />
        </div>
        <div id='stackingLayers'>
          {this.props.layersCRUD.map((layer, index) => {
            return <Layer
              key={index} layerId={index}
              openContext={this.openContext}
              layerDragLeave={this.layerDragLeave}
              layerDragStart={this.layerDragStart}
              layerDragEnd={this.layerDragEnd}
              layerDrop={this.layerDrop}
              layerDragOver={this.layerDragOver}
              layer={layer} />
          })}
        </div>
        <LayerPopupContextMenu hidden={!this.state.displayContext}
          ref={ (ctx) => { this.context = ctx } }
          dismissPopup={this.dismissPopup} />
        <div id='layersButtonsBottom'>
          <button>Group</button>
          <button onClick={() => this.props.addLayer(this.props.layersCRUD.length)}>+</button>
          <button disabled={this.props.layersCRUD.length === 1} onClick={() => this.props.deleteLayer(this.props.changeActiveLayer.layerNumber)}>-</button>
        </div>
      </div>
    )
  }
}

RightPanel.propTypes = {
  changeActiveLayer: PropTypes.object,
  layersCRUD: PropTypes.array,
  changeLayer: PropTypes.func,
  addLayer: PropTypes.func,
  deleteLayer: PropTypes.func,
  reorderLayers: PropTypes.func
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
  return {
    deleteLayer: (layerId) => dispatch(deleteLayer(layerId)),
    addLayer: (layersCount) => dispatch(addLayer(layersCount)),
    changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
    reorderLayers: (oldIndex, newIndex) => dispatch(reorderLayers(oldIndex, newIndex))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(RightPanel);
