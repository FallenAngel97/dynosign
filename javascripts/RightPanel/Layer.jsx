import React from "react";
import "./Layer.scss";
import {connect} from "react-redux";
import {changeActiveLayer, changeLayer, changeLayerVisibility} from "../actions"
import hidden_eye from "./hidden_eye.svg";
import visible_eye from "./visible_eye.svg";

export class Layer extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            editLayer: false
        }
        this.renameLayer = this.renameLayer.bind(this);
        this.handleClick = this.handleClick.bind(this);
        this.layerNameChange = this.layerNameChange.bind(this);
        this.changeVisibility = this.changeVisibility.bind(this);
    }
    componentWillMount() {
        document.addEventListener('mousedown', this.handleClick, false);
    }
    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleClick, false);
    }
    changeVisibility() {
        let layer = this.props.layer;
        this.props.changeLayerVisibility(!layer.hidden, this.props.layerId);
    }
    handleClick(ev) {
        if(!this.node.contains(ev.target) ){
            this.setState({editLayer: false})
        } else if(ev.target.className.indexOf("singleLayer")>-1) {
            this.props._changeActiveLayer(this.props.layer, this.props.layerId);
        }
    }
    renameLayer() {
        this.setState({editLayer: true});
    }
    layerNameChange(ev) {
        let layer = this.props.changeActiveLayer.layer;
        layer.name = ev.target.value;
        this.props.changeLayer(layer, this.props.changeActiveLayer.layerId)
    }
    render() {
        return (
            <div ref={node=>this.node=node} onDoubleClick={this.renameLayer}
                className={"singleLayer"+((this.props.changeActiveLayer.layerNumber === this.props.layerId) ? ' activeLayer':'')}>
                <img onClick={this.changeVisibility} src={this.props.layer.hidden ? hidden_eye : visible_eye}/>
                {this.state.editLayer ? <input onChange={this.layerNameChange} type='text' className='singleLayerEdit' value={this.props.layer.name} /> : this.props.layer.name}
            </div>
        )
    }
}

const mapStateToProps = state => state;

const mapDispatchToProps = dispatch => {
    return {
        _changeActiveLayer: (layer,layerNumber) => dispatch(changeActiveLayer(layer,layerNumber)),
        changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
        changeLayerVisibility: (hidden, layerId) => dispatch(changeLayerVisibility(hidden, layerId))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Layer);