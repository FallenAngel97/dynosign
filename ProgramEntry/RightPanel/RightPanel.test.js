/* global defaultLayer shallow */
import React from 'react';
import { RightPanel, mapDispatchToProps } from './RightPanel';

describe('RightPanel test', () => {
  test('RightPanel can delete layer', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).deleteLayer(0);
    expect(dispatch.mock.calls[0][0]).toEqual(
      { type: 'DELETE_LAYER', layerId: 0 }
    );
    const deleteLayer = jest.fn();
    let rightPanel = shallow(<RightPanel deleteLayer={deleteLayer} layersCRUD={[]} changeActiveLayer={{ layer: { opacity: 0 }, layerNumber: 0 }} />);
    rightPanel.find('#layersButtonsBottom button').at(2).simulate('click');
    expect(deleteLayer).toHaveBeenCalledWith(0);
  })
  test('RightPanel can add layer', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).addLayer(1);
    expect(dispatch.mock.calls[0][0]).toEqual(
      {
        type: 'ADD_LAYER',
        layer:
        {
          ...defaultLayer,
          name: 'New layer ' + 1
        },
        layerNumber: 1
      }
    );
    const addLayer = jest.fn();
    let rightPanel = shallow(<RightPanel addLayer={addLayer} layersCRUD={[]} changeActiveLayer={{ layer: { opacity: 0 }, layerNumber: 0 }} />);
    rightPanel.find('#layersButtonsBottom button').at(1).simulate('click');
    expect(addLayer).toHaveBeenCalledWith(0);
  })
  test('RightPanel can change layer', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).changeLayer(defaultLayer, 0);
    expect(dispatch.mock.calls[0][0]).toEqual(
      { type: 'CHANGE_LAYER', layer: defaultLayer, layerId: 0 }
    );
  });
  test('RightPanel properly renders', () => {
    let rightPanel = shallow(<RightPanel layersCRUD={[]} changeActiveLayer={{ layer: { opacity: 0 } }} />);
    expect(rightPanel).toMatchSnapshot();
    rightPanel = shallow(<RightPanel layersCRUD={[1]} changeActiveLayer={{ layer: { opacity: 0 } }} />);
    expect(rightPanel).toMatchSnapshot();
  });
  test('RightPanel can change opacity', () => {
    const changeLayer = jest.fn();
    let rightPanel = shallow(<RightPanel changeLayer={changeLayer} layersCRUD={[]} changeActiveLayer={{ layer: { opacity: 0 }, layerNumber: 0 }} />);
    rightPanel.find('#topRightPanelIndicators input').simulate('change', { target: { value: 50 } });
    expect(changeLayer).toHaveBeenCalledWith(
      { opacity: 50 }, 0
    );
  });
})
