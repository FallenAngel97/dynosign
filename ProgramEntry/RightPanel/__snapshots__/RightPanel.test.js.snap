// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightPanel test RightPanel properly renders 1`] = `ShallowWrapper {}`;

exports[`RightPanel test RightPanel properly renders 2`] = `ShallowWrapper {}`;
